proxy/src/control/cache.rs

```diff
@@ -93,41 +137,81 @@ impl<T> Cache<T> where T: Clone + Copy + Eq + std::hash::Hash {
         self.reset_on_next_modification = false;
     }
 
-    pub fn clear<F>(&mut self, on_change: &mut F) where F: FnMut(T, CacheChange) {
-        self.retain(&HashSet::new(), on_change)
+    pub fn clear<F>(&mut self, on_change: &mut F)
+    where
+        F: FnMut((K, V), CacheChange),
+    {
+        self.update_intersection(&IndexMap::new(), on_change)
     }
 
-    pub fn retain<F>(&mut self, to_retain: &HashSet<T>, mut on_change: F)
-        where F: FnMut(T, CacheChange)
+    /// Update the cache to contain the intersection of its current contents
+    /// and the key-value pairs in `to_update`. Pairs not present in
+    /// `to_update` will be removed from the cache, and any keys in the cache
+    /// with different values from those in `to_update` will be ovewritten to
+    /// match `to_update`.
+    pub fn update_intersection<F, Q>(
+        &mut self,
+        to_update: &IndexMap<K, V>,
+        mut on_change: F
+    )
+    where
+        F: FnMut((K, V), CacheChange),
+        K: Borrow<Q>,
+        Q: Hash + Eq,
```

Comment:
do we still need the `Q` type? if so, why? -- i think it would be needed if `to_update: &IndexMap<Q, V>`, but that's not the case.

Reply:
Yeah, I think the `Q` type is no longer necessary --- I thought the compiler would warn me about that? Weird.

Reply:
I think it might be better to be generic over borrowed forms of the key here, if we were to eventually use a cache with, e.g., `String` keys, but we don't actually _need_ to for the current use-case.

Reply:
yeah, that's fine if it's actually used by to_update

Reply:
It is as of e860ca46fe555cf560e35e07989e276363a68213